dbms/src/Storages/KVStore/MultiRaft/Disagg/FastAddPeerContext.cpp

```diff
@@ -0,0 +1,174 @@
+}
+
+ParsedCheckpointDataHolderPtr FastAddPeerContext::CheckpointCacheElement::getParsedCheckpointData(Context & context)
+{
+    std::unique_lock lock(mu);
+    if (!parsed_checkpoint_data)
+    {
+        parsed_checkpoint_data = buildParsedCheckpointData(context, manifest_key, dir_seq);
+    }
+    return parsed_checkpoint_data;
+}
+
+std::pair<UInt64, ParsedCheckpointDataHolderPtr> FastAddPeerContext::getNewerCheckpointData(
+    Context & context,
+    UInt64 store_id,
+    UInt64 required_seq)
+{
+    CheckpointCacheElementPtr cache_element = nullptr;
+    UInt64 cache_seq = 0;
+    {
+        std::unique_lock lock{cache_mu};
+        auto iter = checkpoint_cache_map.find(store_id);
+        if (iter != checkpoint_cache_map.end() && (iter->second.first > required_seq))
+        {
+            cache_seq = iter->second.first;
+            cache_element = iter->second.second;
+        }
+    }
+
+    if (!cache_element)
+    {
+        auto s3_client = S3::ClientFactory::instance().sharedTiFlashClient();
+        const auto manifests = S3::CheckpointManifestS3Set::getFromS3(*s3_client, store_id);
+        if (manifests.empty())
+        {
+            LOG_DEBUG(log, "no manifest on this store, skip store_id={}", store_id);
+            return std::make_pair(required_seq, nullptr);
+        }
+        const auto & latest_manifest_key = manifests.latestManifestKey();
+        auto latest_manifest_key_view = S3::S3FilenameView::fromKey(latest_manifest_key);
+        auto latest_upload_seq = latest_manifest_key_view.getUploadSequence();
+        if (latest_upload_seq <= required_seq)
+        {
+            return std::make_pair(required_seq, nullptr);
+        }
+        // check whether there is some other thread downloading the current or newer manifest
+        {
+            std::unique_lock lock{cache_mu};
+            auto iter = checkpoint_cache_map.find(store_id);
+            if (iter != checkpoint_cache_map.end() && (iter->second.first >= latest_upload_seq))
+            {
+                cache_seq = iter->second.first;
+                cache_element = iter->second.second;
+            }
+            else
+            {
+                checkpoint_cache_map.erase(store_id);
+                cache_seq = latest_upload_seq;
+                cache_element = std::make_shared<CheckpointCacheElement>(latest_manifest_key, temp_ps_dir_sequence++);
+                checkpoint_cache_map.emplace(store_id, std::make_pair(latest_upload_seq, cache_element));
+            }
+        }
+    }
+
+    // Now we try to parse the lastest manifest of this store.
+    auto checkpoint_data = cache_element->getParsedCheckpointData(context);
+    return std::make_pair(cache_seq, checkpoint_data);
+}
+
+CheckpointIngestInfoPtr FastAddPeerContext::getOrRestoreCheckpointIngestInfo(
+    TMTContext & tmt,
+    const struct TiFlashRaftProxyHelper * proxy_helper,
+    UInt64 region_id,
+    UInt64 peer_id)
+{
+    std::unique_lock<std::mutex> lock(ingest_info_mu);
```

Comment:
use `std::lock_guard` is enough

Reply:
https://stackoverflow.com/questions/20516773/stdunique-lockstdmutex-or-stdlock-guardstdmutex

Reply:
Originally I want to use a shared_mutex, however, I find mutex is more suitable...